features/address.js

```diff
@@ -0,0 +1,246 @@
+const {chacha20poly1305Encrypt, chacha20poly1305Decrypt, blake2b, sha3_256} = require("./crypto-primitives")
+
+const cbor = require('borc')
+const crc32 = require('../utils/crc32')
+const base58 = require('../utils/base58')
+const bech32 = require('../utils/bech32')
+const pbkdf2 = require('../utils/pbkdf2')
+const variableLengthEncode = require('../utils/variableLengthEncode')
+const CborIndefiniteLengthArray = require('../utils/CborIndefiniteLengthArray')
+const {validateBuffer, validateDerivationScheme, validateArray, validateString, validateNetworkId} = require("../utils/validation")
+
+const AddressTypes = {
+  'BASE': 0b0000,
+  'POINTER': 0b0100,
+  'ENTERPRISE': 0b0110,
+  'BOOTSTRAP': 0b1000,
+  'REWARD': 0b1110
+}
+
+const shelleyAddressTypes = [AddressTypes.BASE, AddressTypes.POINTER, AddressTypes.ENTERPRISE, AddressTypes.REWARD]
+
+const PUB_KEY_LEN = 32
+const KEY_HASH_LEN = 28
+
+function validatePointer(input) {
+  if (!input.hasOwnProperty('blockIndex')
+    || !input.hasOwnProperty('txIndex')
+    || !input.hasOwnProperty('certificateIndex')) {
+    throw new Error('Invalid pointer! Missing one of blockIndex, txIndex, certificateIndex')
+  }
+  if (!Number.isInteger(input.blockIndex)
+    || !Number.isInteger(input.txIndex)
+    || !Number.isInteger(input.certificateIndex)) {
+    throw new Error('Invalid pointer! values must be integer')
+  }
+}
+
+function packBootstrapAddress(derivationPath, xpub, hdPassphrase, derivationScheme) {
```

Comment:
Note: for testnet addresses this should also pack the protocol magic into the address